src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -2430,6 +2429,47 @@ internal void SetFault(Exception ex)
                 this.fault = ex;
             }
 
+            /// <summary>
+            /// Get the correct SynchronizationContext to execute code executing within the lock.
+            /// </summary>
+            private SynchronizationContext GetEffectiveSynchronizationContext()
+            {
+                if (this.synchronizationContext == null)
+                {
+                    // Only read locks can be executed trivially. The locks that have some level of exclusivity (upgradeable read and write)
+                    // must be executed via the NonConcurrentSynchronizationContext.
+                    SynchronizationContext synchronizationContext = null;
+
+                    Awaiter awaiter = this.NestingLock;
+                    while (awaiter != null)
+                    {
+                        if (this.lck.IsLockActive(awaiter, considerStaActive: true))
+                        {
+                            synchronizationContext = awaiter.GetEffectiveSynchronizationContext();
+                            break;
+                        }
+
+                        awaiter = awaiter.NestingLock;
+                    }
+
+                    if (synchronizationContext == null)
+                    {
+                        if (this.kind == LockKind.Read)
+                        {
+                            synchronizationContext = this.lck.defaultSynchronizationContext;
+                        }
+                        else
+                        {
+                            synchronizationContext = new NonConcurrentSynchronizationContext();
+                        }
+                    }
+
+                    Interlocked.CompareExchange(ref this.synchronizationContext, synchronizationContext, null);
```

Comment:
Does `GetEffectiveSynchronizationContext` actually need to be thread-safe? I could be wrong, but it looks like it only runs once per awaiter.

Reply:
I guess we don't need.  I did it because it calls to the function for the nesting lock, so I might not want to depend on the nesting might not even happen, before the parent lock is issued? I guess CompareExchange doesn't have much perf impact.  I can remove it, if you think it might still carry some perf impact?

Reply:
The perf impact is probably insignificant. And if you're not sure it's safe to remove it then by all means, keep it. My primary concern is that when a method uses thread-safe techniques when it is not necessary, it miscommunicates to the code reader who now thinks it *needs* to be thread safe, and then may look for other thread-safety bugs, or assume that it is thread safe and then calls that method in a concurrent way when it in fact never was before and hasn't really been vetted for it.
So I tend to prefer to see a method either 100% thread safe because it needs to be, or not at all, so that it's relatively clear (and ya, you save on unnecessary overhead).

Reply:
I see your point.  Since you have already introduced some metadata to declare whether a method is expected only to be called on the UI thread (or maybe on a specific thread).  Should we consider adding something to declare a method must be thread safe?  Well... maybe it is hard to validate it is thread safe by a validator, but somehow a way to document the behavior of the code/or maybe more important, public contract.  I like the idea to make the code self-documented, even if validators puts it into a limited usage...